### PR TITLE
MLHandler

### PR DESCRIPTION
**Pull request by jaidevd**

### CODE REVIEW

**Comment by jaidevd** [comment]
- Let the blank endpoint (`/mlhandler/`) return a list of currently available models.
- Allow a simpler config:

```yaml
                pattern: /$YAMLURL/mlhandler/   # Only one endpoint per model
                handler: MLHandler
                kwargs:
                        model: $YAMLPATH/iris_lr.pkl
                        xsrf_cookies: false
```

**Comment by jaidevd** [comment]
Add testscases from the reuqests in postman.

**Comment by jaidevd** [comment]
@sanand0 We have only one model per handler now. Please see this sample config:

```yaml
url:
        mlhandler:
                pattern: /$YAMLURL/mlhandler
                handler: MLHandler
                kwargs:
                        # `path` is optional, will be created if it doesn't exist 
                        # from the `class` and `params` keys, as class(**params)
                        # If not specified, will be created under $YAMLPATH as <handler-key>.pkl 
                        path: $YAMLPATH/iris_lr.pkl
                        # optional, will be considered only if `path` doesn't exist
                        model_class: LogisticRegression
                        model_params:
                                C: 10.0
                        xsrf_cookies: false
```
Now, we can do the following:

1. Get predictions from an existing model:
```curl
GET "localhost:9988/mlhandler?sepal_length=5.9&sepal_width=3&petal_length=5.1&petal_width=1.8"
```

2. Retrain existing model on a new target column
```curl
POST 'localhost:9988/mlhandler?_retrain=1&target_col=species' \
--header 'Content-Type: application/json' \
--data-raw '[
    {
        "sepal_length": 5.1,
        "sepal_width": 3.5,
        "petal_length": 1.4,
        "petal_width": 0.2,
        "species": "setosa"
    },
    {
        "sepal_length": 4.9,
        "sepal_width": 3.0,
        "petal_length": 1.4,
        "petal_width": 0.2,
        "species": "setosa"
    }, # etc
]'
```

3. Get bulk predictions
```curl
POST 'localhost:9988/mlhandler' \
--header 'Content-Type: application/json' \
--data-raw (see above)
```

4. See model parameters and last training score
```curl
GET 'localhost:9988/mlhandler?_model'
```

5. Create a model with an sklearn class
```curl
PUT 'localhost:9988/mlhandler?class=GaussianNB'
```

6. Change parameters of an existing model
```curl
PUT 'localhost:9988/mlhandler?class=LogisticRegression&class_weight=balanced&C=100.0'
```

7. Change parameters and retrain model
```curl
PUT 'localhost:9988/mlhandler?class=RandomForestClassifier&_retrain=1&target_col=species&n_estimators=1' \
--header 'Content-Type: application/json' \
--data-raw '[see above]'
```

8. Delete a model
```curl
DELETE 'localhost:9988/mlhandler'
```

I'll convert the above into tests.

**Comment by jaidevd** [comment]
What happens when model_class doesn't exist, but we start with a PUT?

**Comment by jaidevd** [comment]
Try out examples of:
- forecasting
- classification

**Comment by sanand0** [comment]
@jaidevd -- I was playing with this. Two immediate issues:

First, `model_class` should be optional. If I just specify a `path: model.pkl`, it still searches for model_class

```
Traceback (most recent call last):
  File "d:\site\gramener.com\viz\async-gramex\gramex\services\__init__.py", line 720, in url
    urlspec.handler.setup(**kwargs)
  File "d:\site\gramener.com\viz\async-gramex\gramex\handlers\mlhandler.py", line 79, in setup
    mclass = search_modelclass(model_class)
  File "d:\site\gramener.com\viz\async-gramex\gramex\handlers\mlhandler.py", line 43, in search_modelclass
    cls = pydoc.locate(mclass)
  File "D:\anaconda\3.7\lib\pydoc.py", line 1605, in locate
    parts = [part for part in path.split('.') if part]
AttributeError: 'NoneType' object has no attribute 'split'
```

Second, I was not able to figure out how to POST training data via a file via curl. How do I do that?

**Comment by jaidevd** [comment]
@sanand0 I'm not able to reproduce this. If I have a `model.pkl`, say, trained on iris, and the following config:
```yaml
url:
        model:
                pattern: /$YAMLURL/ml
                handler: MLHandler
                kwargs:
                        path: $YAMLPATH/model.pkl
                        xsrf_cookies: false
```
then, the following works:

```
GET /ml?sepal_length=5.9&sepal_width=3&petal_length=5.1&petal_width=1.8
```

Posting files with curl is done as follows:

```
curl -X POST -F 'file=@/tmp/iris.csv' http://localhost:9988/ml
```

This ^ gets predictions for all rows in `/tmp/iris.csv`.

**Comment by jaidevd** [comment]
Caching and concatenating data should be optional. There should be a way to enforce no caching - especially for public interfaces like the Gramex guide.

**Comment by jaidevd** [comment]
make the inline example as forms, inline

**Comment by jaidevd** [comment]
Allow a complete cold start - no existing models, data, or model classes.

**Comment by sanand0** [comment]
I tested with [Home credit default risk](https://www.kaggle.com/c/home-credit-default-risk/data?select=application_train.csv).

- [ ] Can we have a handler without a path and a model as well? Just MLHandler!
- [ ] Note: when creating the path, `slugify(cls.name)`
- [ ] If path is not specified, where should it be saved -- keeping portability in mind?

```yaml
url:
  neighbors:
    pattern: /neighbors
    handler: MLHandler
    kwargs:
      xsrf_cookies: false
```

- [ ] `curl /neighbors` -> more intuitive error message (e.g. "model is not created")
- [ ] `curl /neighbors?_model` -> more intuitive error message

```yaml
url:
  neighbors:
    pattern: /neighbors
    handler: MLHandler
    kwargs:
      model_class: DecisionTreeClassifier
      xsrf_cookies: false
```

- [x] `curl 'localhost:9988/neighbors?_model&class=KNeighborsClassifier'` works
- [ ] `curl 'localhost:9988/neighbors?_model'` doesn't tell me what model we're using
- [ ] I was thrown away by `_retrain=` beginning with `_` but `target_col` not doing so
- [ ] I took the first 1000 rows into 1000.csv
      `curl -X POST -F _retrain=1 -F target_col=TARGET -F 'data=@1000.csv' 'localhost:9988/neighbors'`
      "ValueError: could not convert string to float: 'Cash loans'". How to auto-label-encode?
- [ ] I encoded / removed string columns into 1000a.csv
      `curl -X POST -F _retrain=1 -F target_col=TARGET -F 'data=@1000a.csv' 'localhost:9988/neighbors'`
      "ValueError: Input contains NaN, infinity or a value too large for dtype('float32')". How to specify pipelines?
- [x] I removed columns with missing values into 1000b.csv
      `curl -X POST -F _retrain=1 -F target_col=TARGET -F 'data=@1000b.csv' 'localhost:9988/neighbors'`
- [ ] Can we pretty-print JSON output from `?_model`? It's hard to read on console / browser
- [ ] How do I get the list of input columns to pass to the model?
- [x] `curl 'localhost:9988/neighbors?_model&class=DecisionTreeClassifier'` works
- [x] I added 10 rows with target_col=0 and 1 respectively into 1000b-test.csv
      `curl -X POST -F 'data=@1000b-test.csv' 'localhost:9988/neighbors'`
- [ ] Can I incrementally classify? This will be of great help

Top next steps:

- Guessable API
- Friendly output and error messages
- Pipelines
- Incremental training

**Comment by jaidevd** [comment]
- [ ] For errors, instead of sending HTML responses, send readable JSON responses.
- [ ] @sanand0 : Create a JSONResponse template, and make this the default for MLHandler in gramex/gramex.yaml
- [ ] Start sending / receiving data column names in GET, PUT, POST, etc.
- [ ] `transorm:` keys for transforming data.


```
GET /mlhandler?.._target_col=Species&_ohe=sepal_width
```

**Comment by jaidevd** [comment]
Apparently there is no straightforward way to do [partial fitting](https://stackoverflow.com/questions/17931865/using-partial-fit-with-scikit-pipeline) with [pipelines](https://github.com/scikit-learn/scikit-learn/issues/3299). This can be done by hacking the pipeline class. Let's evaluate when we need it.

**Comment by jaidevd** [comment]
- Single endpoint, multiple models, so `self.model` doesn't have to exist.
- Support partial_fit if it exists, by default. If the user has to do a full fit, ask them to delete the data cache.

**Comment by jaidevd** [comment]
```
url:
        mlhandler:
                pattern: /$YAMLURL/ml
                handler: MLHandler
                kwargs:
                        data: $YAMLPATH/train.csv
                        model:
                                class: LogisticRegression
                                exclude: [PassengerId, Ticket, Cabin, Name]
                                target_col: Survived
                                pipeline: true
                                nums: [Age, Fare, SibSp, Parch]
                                cats: [Pclass, Sex, Embarked]
```

**Comment by sanand0** [comment]
- [ ] Make `pipeline=true` the default
- [ ] Ensure `cats=` and `nums=` are added to `include=` if they're not already there
- [ ] POST should decode self.request.body as JSON only if Content-Type is application/json -- just like FormHandler
- [ ] We get a `AttributeError: 'MLHandler' object has no attribute '_data_cachedir'` on `curl -X POST -F'file=@head-test.csv' localhost:9988/model`
- [ ] Model training & transformation must be async

**Comment by sanand0** [comment]
- [ ] Save models at `GRAMEXDATA/apps/mlhandler/*`
- [ ] GET `/model` should show a rendered template that users can override via `template: <path|false>`. If it's false, return an empty array. Else render the template
- [ ] GET `/model?_model` should return the model parameters -- the value doesn't need to be truthy
- [ ] `/model/` should set self.set_header('Content-Type', 'application/json')
- [ ] When `gramex.yaml` changes and `gramex.init()` is re-run, MLHandler raises a thread exception
- [ ] "Model saved..." message should read "{self.name}: model at {path}"
- [ ] After we delete the model, it returns a 500. It should return a 404 (or any 40x)
- [ ] After we delete the model and send a `POST -F... /model` it raises a KeyError: 'TARGET'. It should pick up the model class, etc from the last model parameters set -- irrespective of whether these came from YAML or from a `PUT ?_model&....`
- [ ] Should GET `/model?_model` return the model quality EVEN THOUGH sklearn doesn't?

**Comment by sanand0** [comment]
- [ ] On a missing model, `/?_model` returns a HTML error. It should return a JSON error
- If YAML has `model.class: DecisionTreeClassifier` and nothing else, then
  - [ ] `PUT ?_model&target_col=TARGET` fails at `mclass = self.args.pop('class')[0]`. It should pick up the class (or ANY PRIOR CONFIG) from the YAML, or the most recent PUT
  - [ ] `PUT '?_model&target_col=TARGET&class=DecisionTreeClassifier'` fails at `return {k: param_types[k](v) for k, v in new_params.items()}` KeyError: target_col
  - [ ] `POST` on both these returns a 404
  - [ ] Same for `cats` and `nums`
  - [ ] `POST` a single row returns a 404 -- with or without `?_action=append`
- [ ] If we train a model with one-hot encoding that only has gender=M, the next time it gets a gender=F it fails

**Comment by sanand0** [comment]
The following sequence raises an error

- Create a model with no kwargs except `xsrf_cookies: false`
- Setup model: `curl -X PUT 'localhost:9988/model?_model&class=DecisionTreeClassifier&target_col=TARGET'`
- Train: `curl -X POST -F 'file=@line.csv' 'localhost:9988/model?_action=train&target_col'`
- Fetch model: `curl localhost:9988/model`

This reports `FileNotFoundError: File C:\Users\anand\AppData\Local\Gramex Data\apps\mlhandler\mlhandler-loans-model\data.h5 does not exist`


- [ ] I should be able to send `?_action=append&_action=retrain` which will append and retrain
- [ ] If YAML changes, don't retrain the model. Instead
  - set `trained=false` to model parameters
  - when trained is False, `app_log.warn()` asking them to run `$YAMLURL/?action=retrain`
  - TODO: Evaluate whether we send a HTTP header X-Model-Trained: 0

**Comment by jaidevd** [comment]
- [ ] When dropping NAs and duplicates, ignore columns that have been excluded.
- [ ] test with existing models that don't follow MlHandler's default way of creating pipelines. Pick something completely different from what MlHandler usually creates, and try crudding it. - - [ ] When loading existing models, populate the config store if it's empty, with whatever can be inferred from the model / pipeline. Some pipeline components may have a `feature_names` attribute as well.

**Comment by jaidevd** [comment]
- [ ] Support `?_action=score` with scoring metrics available here: https://scikit-learn.org/stable/modules/model_evaluation.html#common-cases-predefined-values

File: gramex/apps/mlhandler/template.html [hunk 1]
```diff
@@ -0,0 +1,57 @@
+<!DOCTYPE html>
```
Comment (jaidevd) [inline gramex/apps/mlhandler/template.html:1]: Swap the order of the sections - training data, then model.

**Pull request merged**